### DISCUSSION TITLE
Lofting between two faces

### DISCUSSION THREAD

**Discussion by napter** [elalish/manifold#410] [Q&A]
Is there a way with manifold to triangulate between two faces with differing numbers of nodes to create a 3d object? Say a square and a circle on different planes?

**Comment by elalish** [comment]
Not yet, though I'd like to play with that someday. It's a very open-ended problem without a clear "right way". I was always told with CAD programs you pay for the lofting (as in that's the hardest part to do well).

**Reply by pca006132** [reply]
This somehow makes me think of my first PR, trying to add a sweep function. Maybe loft between faces with differing number of vertices can be done by adding vertices to both faces to make each edge equal distance? Although this only work for simple polygons, and is probably not the best way of implementing this. (I know nothing about geometry lol)

**Reply by geoffder** [reply]
This operation is called skin in [BOSL2](https://github.com/revarbat/BOSL2/wiki/skin.scad), which comes with a few different algorithms for matching up the points between the input shapes (sorting + resampling, or inserting points based on distance or tangent matching). One of the trickier bits of that library that I've ported it in OCADml ([skin](https://github.com/OCADml/OCADml/blob/main/lib/skin.ml) which creates the mesh, and the [matching helpers](https://github.com/OCADml/OCADml/blob/main/lib/pathMatch.ml)).

I replicated the vacuum example they have in BOSL2 for a basic usage [demo](https://ocadml.github.io/OSCADml/OSCADml/profile_skinning.html), but I also made a morphing sweep function that reuses the skinning helpers to be able to make things like [these](https://ocadml.github.io/OSCADml/OSCADml/morphing_sweeps.html) a bit easier.

**Reply by napter** [reply]
@geoffder the Mesh.morphing_sweep examples are excellent, what would it take to implement that with manifold?

**Reply by napter** [reply]
> This somehow makes me think of my first PR, trying to add a sweep function. Maybe loft between faces with differing number of vertices can be done by adding vertices to both faces to make each edge equal distance? Although this only work for simple polygons, and is probably not the best way of implementing this. (I know nothing about geometry lol)

Is it possible to do a sweep with manifold (js library) currently? I saw that PR but I don't think it was ever merged.

**Reply by elalish** [reply]
We didn't merge sweep because it can be done with `Warp`, which is lower-level. I imagine we'll want a sweep function in manifold eventually, but I'd rather wait until we have a solution for self-intersections, since they are created so easily with a sweep.

**Reply by geoffder** [reply]
> self-intersections, since they are created so easily with a sweep

They are indeed. I've used a simple check for points of the next profile being "above" the plane of the previous profile by some epsilon in troublesome parts of my dometyl keyboard generator to try and *avoid* self intersections in tight sweeps with mixed success. I'm no geometrician so I'm not aware of there is a more reliable automatic avoidance scheme or post-fixing algorithm already existing out there.

**Reply by napter** [reply]
> We didn't merge sweep because it can be done with `Warp`, which is lower-level. I imagine we'll want a sweep function in manifold eventually, but I'd rather wait until we have a solution for self-intersections, since they are created so easily with a sweep.

So it can be done but will not create a manifold shape because it will have self-intersections. Can it be done by making each new section using an extrude + warp, then unioning the new section with the growing shape? Is this possible today and would it be performant? Is there a better way to do a sweep?

**Reply by elalish** [reply]
Technically speaking, manifoldness (topology) and overlaps (geometry) are separate concepts. Yes, a bunch of unions is possible, no it's not that performant, but it'll probably work. I'd still say Warp is a better way to go, but you have to be careful to keep it from curling too sharply. See my knot example. 

#289 is what I want to do to handle overlaps properly (this library is robust to them, but you'll still have some surprising results sometimes). That'll be a fair bit of work though.

**Reply by napter** [reply]
What I'd like to do is sweep along an arbitrary path that may curl sharply. Any recommendation of how to do that, with or without warp, without ending up with overlapping geometry (or some way of cleaning them, prior to the implementation of #289)?

**Reply by elalish** [reply]
Yeah, for now a stack of unions is probably the way to go. It'll still probably give some ugly results when you curl sharply though, but at least it won't overlap itself.

**Comment by cartesian-theatrics** [comment]
I wrote a simple N-segment Loft algorithm for isomorphic cross-sections. It's very basic, but even as such I find it to be quite useful. It takes either a single cross-section (or vector of Polygons) and N transforms, or N sections and N transforms, and lofts between the transformed cross-sections 1-to-1. In the future I'll probably extend it to support various non-isomorphic joins.


```cpp

manifold::Manifold Loft(const std::vector<manifold::Polygons>& sections, const std::vector<glm::mat4x3>& transforms) {
    std::vector<glm::vec3> vertPos;
    std::vector<glm::ivec3> triVerts;

    if (sections.size() != transforms.size()) {
        throw std::runtime_error("Mismatched number of sections and transforms");
    }

    std::size_t offset = 0;
    std::size_t nVerticesInEachSection = 0;

    for (std::size_t i = 0; i < sections.size(); ++i) {
        const manifold::Polygons polygons = sections[i];
        glm::mat4x3 transform = transforms[i];

        for (const auto& polygon : polygons) {
            for (const glm::vec2& vertex : polygon) {
                glm::vec3 translatedVertex = MatrixTransforms::Translate(transform, glm::vec3(vertex.x, vertex.y, 0))[3];
                vertPos.push_back(translatedVertex);
            }
        }

        if (i == 0) {
            nVerticesInEachSection = vertPos.size();
        } else if ((vertPos.size() % nVerticesInEachSection) != 0)  {
            throw std::runtime_error("Recieved CrossSection with different number of vertices");
        }

        if (i < sections.size() - 1) {
            std::size_t currentOffset = offset;
            std::size_t nextOffset = offset + nVerticesInEachSection;

            for (std::size_t j = 0; j < polygons.size(); ++j) {
                const auto& polygon = polygons[j];

                for (std::size_t k = 0; k < polygon.size(); ++k) {
                    std::size_t nextIndex = (k + 1) % polygon.size();

                    glm::ivec3 triangle1(currentOffset + k, currentOffset + nextIndex, nextOffset + k);
                    glm::ivec3 triangle2(currentOffset + nextIndex, nextOffset + nextIndex, nextOffset + k);

                    triVerts.push_back(triangle1);
                    triVerts.push_back(triangle2);
                }
                currentOffset += polygon.size();
                nextOffset += polygon.size();
            }
        }

        offset += nVerticesInEachSection;
    }

    auto frontPolygons = sections.front();
    auto frontTriangles = manifold::Triangulate(frontPolygons, -1.0);
    for (auto& tri : frontTriangles) {
        triVerts.push_back({tri.z, tri.y, tri.x});
    }

    auto backPolygons = sections.back();
    auto backTriangles = manifold::Triangulate(backPolygons, -1.0);
    for (auto& triangle : backTriangles) {
        triangle.x += offset - nVerticesInEachSection;
        triangle.y += offset - nVerticesInEachSection;
        triangle.z += offset - nVerticesInEachSection;
        triVerts.push_back(triangle);
    }

    manifold::Mesh mesh;
    mesh.triVerts = triVerts;
    mesh.vertPos = vertPos;
    return manifold::Manifold(mesh);
}

manifold::Manifold Loft(const manifold::Polygons& sections, const std::vector<glm::mat4x3>& transforms) {
    std::vector<manifold::Polygons> polys;
    for (auto section : sections) {
        polys.push_back({section});
    }
    return Loft(polys, transforms);
}

manifold::Manifold Loft(const manifold::SimplePolygon& section, const std::vector<glm::mat4x3>& transforms) {
    std::vector<manifold::Polygons> polys;
    for (std::size_t i = 0; i < transforms.size(); i++) {
        polys.push_back({section});
    }
    return Loft(polys, transforms);
}

manifold::Manifold Loft(const std::vector<manifold::CrossSection>& sections, const std::vector<glm::mat4x3>& transforms) {
    std::vector<manifold::Polygons> polys;
    for (auto section : sections) {
        polys.push_back(section.ToPolygons());
    }
    return Loft(polys, transforms);
}

manifold::Manifold Loft(const manifold::CrossSection section, const std::vector<glm::mat4x3>& transforms) {
    std::vector<manifold::Polygons> sections(transforms.size());
    auto polys = section.ToPolygons();
    for (std::size_t i = 0; i < transforms.size(); i++) {
        sections[i] = polys;
    }
    return Loft(sections, transforms);
}
```

The biggest issue I encounter is that the filtering of vertices in cross-sections can be ambiguous (it seems to remove coplananr vertices), resulting in high strangeness at times. That's why it also supports just plain polygons.

This implementation does require that you preserve unit length of columns in the transform matrix. I use Rodriguez rotations for this and I have a set of operations for manipulating transforms in place or "egocentrically". This makes it really easy to define "paths" of transform frames.

```cpp

glm::vec3 rodrigues_rotation(const glm::vec3& v, const glm::vec3& k, float a) {
    return v * glm::cos(a) + glm::cross(k, v) * glm::sin(a) + k * glm::dot(k, v) * (1 - glm::cos(a));
}

glm::mat4x3 Yaw(const glm::mat4x3& m, float a) {
    glm::vec3 d[] = {
        rodrigues_rotation(m[0], m[1], a),
        m[1],
        rodrigues_rotation(m[2], m[1], a)
    };
    return glm::mat4x3(d[0], d[1], d[2], m[3]);
}

glm::mat4x3 Pitch(const glm::mat4x3& m, float a) {
    glm::vec3 d[] = {
        m[0],
        rodrigues_rotation(m[1], m[0], a),
        rodrigues_rotation(m[2], m[0], a)
    };
    return glm::mat4x3(d[0], d[1], d[2], m[3]);
}

glm::mat4x3 Roll(const glm::mat4x3& m, float a) {
    glm::vec3 d[] = {
        rodrigues_rotation(m[0], m[2], a),
        rodrigues_rotation(m[1], m[2], a),
        m[2]
    };
    return glm::mat4x3(d[0], d[1], d[2], m[3]);
}

glm::mat4x3 Rotate(const glm::mat4x3& m, const glm::vec3& axis, float a) {
    glm::vec3 d[] = {
        rodrigues_rotation(m[0], axis, a),
        rodrigues_rotation(m[1], axis, a),
        rodrigues_rotation(m[2], axis, a)
    };
    return glm::mat4x3(d[0], d[1], d[2], m[3]);
}

glm::mat4x3 InvertRotate(const glm::mat4x3& m, const glm::vec3& angles) {
    glm::mat4x3 res = m;
    if (angles[2] != 0) {
        res = Roll(res, -angles[2]);
    }
    if (angles[1] != 0) {
        res = Yaw(res, -angles[1]);
    }
    if (angles[0] != 0) {
        res = Pitch(res, -angles[0]);
    }
    return res;
}

glm::mat4x3 Rotate(const glm::mat4x3& m, const glm::vec3& angles) {
    glm::mat4x3 res = m;
    if (angles[0] != 0) {
        res = Pitch(res, angles[0]);
    }
    if (angles[1] != 0) {
        res = Yaw(res, angles[1]);
    }
    if (angles[2] != 0) {
        res = Roll(res, angles[2]);
    }
    return res;
}

glm::mat4x3 SetRotation(const glm::mat4x3& m, const glm::mat3x3& rotation) {
    glm::mat4x3 result = m;
    for (int i = 0; i < 3; ++i) {
        result[i] = rotation[i];
    }
    return result;
}

glm::mat4x3 Translate(const glm::mat4x3& m, const glm::vec3& offset) {
    glm::mat4x3 result = m;

    if (offset[0] != 0.0) {
        result[3] += m[0]*offset[0];
    }
    if (offset[1] != 0.0) {
        result[3] += m[1]*offset[1];
    }
    if (offset[2] != 0.0) {
        result[3] += m[2]*offset[2];
    }

    return result;
}

glm::mat4x3 SetTranslation(const glm::mat4x3& m, const glm::vec3& translation) {
    glm::mat4x3 result = m;
    result[3] = translation;
    return result;
}

glm::mat4x3 InvertTransform(const glm::mat4x3& m) {

  glm::mat3x3 rotationPart = glm::mat3x3(m);
  glm::mat3x3 unRotated = SetRotation(m, glm::transpose(rotationPart));

  glm::vec3 translation = -m[3];
  glm::mat4x3 result = Translate(unRotated, translation);

  return result;
}

glm::mat4x3 CombineTransforms(const glm::mat4x3& a, const glm::mat4x3& b) {
    glm::mat4x3 result;

    // Apply rotation of 'a' to 'b' and normalize
    for (int i = 0; i < 3; ++i) {
        result[i] = glm::normalize(a[0]*b[i][0] + a[1]*b[i][1] + a[2]*b[i][2]);
    }

    // Apply translation of 'a' to 'b'
    result[3] = a[0]*b[3][0] + a[1]*b[3][1] + a[2]*b[3][2] + a[3];

    return result;
}
```

**Comment by geoffder** [comment]
Supporting skinning over a variety of polygon profiles nicely (imo) requires providing algorithms for mapping between vertices of shapes with differing numbers of points, and even then ugly and self-intersecting meshes are quite easy to create. I'm not sure whether this sort of thing falls under what manifold is trying to cover, but I could be misinterpreting @elalish.

**Reply by cartesian-theatrics** [reply]
Yes it's hard to make something that's valid for all possible and/or reasonable inputs, but the basic idea of what's going on is easy enough to understand. In the future I'd like to at least have a "nearest neighbor" JoinType.

**Reply by elalish** [reply]
Yes, the old adage I always heard was in CAD, you pay for the lofting. The idea being it is such a large, difficult, and ambiguous problem that a lot of work goes into the various algorithms. For now I'm happy to have people building lofting algorithms on top of Manifold, though I'm sure eventually certain ones will come into the core codebase. It would be great to see some examples of objects you've created with this loft function.

From a code simplicity perspective, I would probably rewrite this by using `Extrude` on the first polygon, then `Warp` to morph one cross-section into the other.

**Reply by cartesian-theatrics** [reply]
For sure, this is definitely not an adequate implementation as is. It might be interesting to take a more explicit approach with a parameterized skinning algorithm like JointType for offset, especially since the time complexity may vary substantially depending on the algorithm.

I'll often use it for f(z) type constructions, like for example with this spiralized hydroponic tower segment
![LoftExample](https://github.com/elalish/manifold/assets/26554374/4dc57dc4-480a-4916-b859-dce05c947436)

Or I used it for two-part mold for a static mixer:
![PXL_20221014_080513189](https://github.com/elalish/manifold/assets/26554374/4ead76f4-2022-49a1-b6aa-daf697ba714e)

If you're planning ahead of time, it's usually easy to ensure f(z) produces a polygon with N vertices. In cases where I need a dimensionality change I'll often use convex hulls instead.

I also use it for things like pipes with lots of turns and bends. There are often slightly annoying edge cases when unioning linear segments end-to-end.
